data_store/elastic.go

```
@@ -8,87 +8,114 @@ import (
 	"github.com/olivere/elastic/v7"
 )
 
-const INDEX_NAME = "books_yuval"
-const PORT = 9200
-const DOMAIN_URL = "http://es-search-7.fiverrdev.com"
+const (
+	INDEX_NAME = "books_yuval"
+	PORT       = 9200
+	DOMAIN_URL = "http://es-search-7.fiverrdev.com"
+)
 
 type Elastic struct {
 	Client *elastic.Client
 }
 
-func (e *Elastic) InitClient() error {
-	setURL := fmt.Sprint(DOMAIN_URL, ":", PORT)
+func initElasticClient() *elastic.Client {
+	setURL := fmt.Sprintf("%s:%d", DOMAIN_URL, PORT)
 	client, err := elastic.NewClient(elastic.SetURL(setURL))
 	if err != nil {
-		return err
+		return nil
 	}
-	e.Client = client
-	return nil
+
+	return client
+}
+
+func CreateElastic() *Elastic {
+	e := Elastic{Client: initElasticClient()}
+	if e.Client == nil {
+		return nil
+	}
+
+	return &e
 }
```

```suggestion
func newElasticClient() *Elastic {
	client :=  initElasticClient()
	if client != nil {
		return nil
	}

	return &Elastic { client: client}
}
```

---

and I think you should return an error if needed from `initElasticClient` function

---

if initElasticClient returns an error then the caller, newElasticClient, should also return the error up to NewBookStorer right?
I remember you told me such constructors normally don't return error just nil or value.
which is it? not sure what you mean...

---

you are right - the contractor will return nil - and the `initElasticClient` (private function) will return the err - and sometimes it is fine to return an err as a second value ...